File: 02-primes/primes.java
class Primes {
	static int[] primes;

	static boolean isPrime(int x) {
		for (int d : primes) {
			if (d * d > x) return true;
			if (x % d == 0) return false;
		}
		throw new RuntimeException("never should get here");
	}
	
	public static void main(String[] argv) {
		int maxN = Integer.parseInt(System.getenv("MAXN"));
		primes = new int[maxN];
		primes[0] = 2;
		primes[1] = 3;
		primes[2] = 5;
		primes[3] = 7;
		int c = 4;
		for (int i = 9; true; i += 2) {
			if (isPrime(i)) {
				primes[c++] = i;
				if (c == maxN) {
					System.out.printf("primes[%d] = %d%n",
						maxN, primes[maxN - 1]);
					break;
				}
			}
		}
	}
}

File: 02-primes/primes.erl
-module(primes).
-export([start/0]).

divfound(P, I, A) ->
  D = array:get(I, P),
  if
    D * D > A -> false;
    A rem D == 0 -> true;
    true -> divfound(P, I+1, A)
  end.

primes(N, P, A) ->
  S = array:size(P),
  if
    N == S -> P;
    true -> case divfound(P, 0, A) of
      true -> primes(N, P, A+2);
      false -> primes(N, array:set(array:size(P), A, P), A+2)
    end
  end.

start() ->
  N = list_to_integer(os:getenv("MAXN")),
  io:format("primes[~b] = ~b~n", [N, array:get(N-1, primes(N, array:from_list([2, 3, 5, 7]), 9))]).
